### PR TITLE
[sparsity] Enable FX quantization for sparse linear

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66779
* #66778
* #66777
* #69415

Tests: TBD

Differential Revision: [D31835720](https://our.internmc.facebook.com/intern/diff/D31835720)

### CODE REVIEW

**Review by vkuzo — CHANGES_REQUESTED** [review]
It's not ideal to attach a custom attribute to a module in one framework and then check for that attribute in a different framework.  A way to make this more discoverable would be to annotate the `sparse_params` attribute on all types of modules which are expected to support it.  Requesting changes for now, would be great to discuss next steps.

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31835720).

File: torch/ao/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -955,15 +956,23 @@ def convert(self,
...
                 if activation_int8_quantized:
                     additional_static_quant_mapping = convert_custom_config_dict.get("static", {})
-                    qlinear = get_static_quant_module_class(
-                        type(self.linear), additional_static_quant_mapping)
+                    if module_has_sparse_params(self.linear):
+                        mapping_fn = get_static_sparse_quant_module_class
+                    else:
+                        mapping_fn = get_static_quant_module_class
+                    qlinear = mapping_fn(type(self.linear),
+                                         additional_static_quant_mapping)
```
Comment (vkuzo) [inline torch/ao/quantization/fx/quantization_patterns.py:964]: this still means that quantization knows about sparsity.  If we want them to be fully independent, we could move this to an external API.  Something like

```
prepare_custom_config_dict = {
  'float_to_observed_custom_module_class_with_lambda': {
    'static': {
      'float_class': torch.nn.Linear,
      'observed_class': torch.ao.nn.sparse.Linear,  # or whatever the type is
      'lambda': lambda x: module_has_sparse_params(x),
    },
  },
  # repeat for observed to quantized
}
```

and then make the quantization prepare and convert honor these swaps with lambda

or something similar

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31835720).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31835720).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31835720).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31835720).

**Pull request closed without merging**